beginners-guide.md

```diff
@@ -19,10 +19,18 @@ SIMD has a few special vocabulary terms you should know:
 
+* **Vectorize**: An operation that uses SIMD instructions to operate over a vector is often referred to as "vectorized".
+
+* **Autovectorization**: Also known as _implicit vectorization_. This is when a compiler can automatically recognize a situation where scalar instructions may be replaced with SIMD instructions, and use those instead. A compiler lacks the same knowledge about the purpose of the program and shape of the data as the programmer who is using it, and that kind of domain knowledge strongly affects when and where SIMD instructions are appropriate, so this compiler optimization can often fail to trigger, even when it seems obvious to the programmer.
+
+* **Scalar:** "Scalar" in mathematical contexts refers to values that have only one "dimension", like real numbers (which can be represented by a line), but in programming it can also be used to describe "scalar operations" that use strictly scalar values, like addition. This term is mostly used to differentiate between vectorized operations that use SIMD instructions versus scalar operations that don't.
+
 * **Lane:** A single element position within a vector is called a lane. If you have `N` lanes available then they're numbered from `0` to `N-1` when referring to them, again like an array. The biggest difference between an array element and a vector lane is that in general is *relatively costly* to access an individual lane value. On most architectures, the vector has to be pushed out of the SIMD register onto the stack, then an individual lane is accessed while it's on the stack (and possibly the stack value is read back into a register). For this reason, when working with SIMD you should avoid reading or writing the value of an individual lane during hot loops.
 
 * **Bit Widths:** When talking about SIMD, the bit widths used are the bit size of the vectors involved, *not* the individual elements. So "128-bit SIMD" has 128-bit vectors, and that might be `f32x4`, `i32x4`, `i16x8`, or other variations. While 128-bit SIMD is the most common, there's also 64-bit, 256-bit, and even 512-bit on the newest CPUs.
 
+* **Vector Register:** The extra-wide registers that are used for SIMD operations are commonly called vector registers, though you may also see "SIMD registers", vendor names for specific features, or even "floating-point register" as it is common for the same registers to be used with both scalar and vectorized floating-point operations. A processor will execute a SIMD instruction with a vector register, but it may not be executed on a vector register with the same width as the vector that is actually being operated on, because an ISA may not have the same registers available on each processor, and a processor may be able to activate or deactivate registers at runtime. This can affect the performance of SIMD code.
```

Comment:
I'd only add something about "They look like `xmm`, `ymm` or `zmm` when you look at assembly, and that's a way to know you are using the vector registers. Good addition.

Reply:
There's different names for different architectures, but I think it would be totally reasonable to give the names for NEON, SSE, and AVX{,2,512} registers since those are all the ones in tier1 support.

Reply:
I think everything starting at "a processor will execute" and after is getting deep into the weeds for a glossary.

Reply:
Also true, I think probably this should be discussed somewhere but moved into another section, thinking about it. We want the glossary to be (relatively) laconic so it's quickly reviewed.